File: src/linear_algebra/C/vectors.c
#include "vectors.h"

#include "omp.h"
#include <math.h>


double norm(int n, double v[]) {
    double sum = 0;

    #pragma omp parallel for reduction(+:sum)
    for (int i = 0; i < n; i++) {
        sum += v[i] * v[i];
    }

    return (double) sqrt(sum);
}

/* Vector Addition */
void v_add(double u[], double v[], int n, double result[]) {    
    for (int i = 0; i < n; i++) {
        result[i] = u[i] + v[i];
    }
}

/* Vector Subtraction */
void v_sub(double u[], double v[], int n, double result[]) {
    for (int i = 0; i < n; i++) {
        result[i] = u[i] - v[i];
    }
}

/* Vector Multiplication by Scalar (serial) */
void v_mult_s(double v[], double c, int n, double result[]) {
    for (int i = 0; i < n; i++) {
        result[i] = v[i] * c;
    }
}

/* Vector Multiplication by Scalar (parallel) */
void v_mult_p(double v[], double c, int n, double result[]) {
    int i;
    #pragma omp parallel for
    for (i = 0; i < n; i++) {
        result[i] = v[i] * c;
    }
}

/* Vector Dot Product */
double dot(double u[], double v[], int n) {
    double sum = 0;
    for (int i = 0; i < n; i++) {
        sum += u[i] * v[i];
    }

    return sum;
}

/* Vector Cross Product */
void cross(double u[], double v[], int n, double result[]) {
    if (n != 3) {
        return;
    }
    
    result[0] = u[1] * v[2] - u[2] * v[1];
    result[1] = u[2] * v[0] - u[0] * v[2];
    result[2] = u[0] * v[1] - u[1] * v[0];
}

/* Triple Vector Product */
void triple(double u[], double v[], double w[], int n, double result[]) {
    if (n != 3) {
        return;
    }
    
    double temp[3];
    cross(v, w, n, temp);
    cross(u, temp, n, result);
}

void v_hadamard_s(double u[], double v[], int n, double result[]) {
    for (int i = 0; i < n; i++) {
        result[i] = u[i] * v[i];
    }
}

void v_hadamard_p(double u[], double v[], int n, double result[]) {
    int i;
    #pragma omp parallel for
    for (i = 0; i < n; i++) {
        result[i] = u[i] * v[i];
    }
}

void v_outer_s(double u[], double v[], int n, double result[n][n]) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            result[i][j] = u[i] * v[j];
        }
    }
}

void v_outer_p(double u[], double v[], int n, double result[n][n]) {
    int i, j;
    #pragma omp parallel for collapse(2)
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            result[i][j] = u[i] * v[j];
        }
    }
}

File: src/linear_algebra/vectors.py
import numpy as np

def add(u, v):
    if u.shape != v.shape:
        print("ERROR: vector lengths don't match\n")
        return None
    
    n = u.shape[0]
    result = np.zeros(n)
    for i in range(n):
        result[i] = u[i] + v[i]
    return result

def sub(u, v):
    if u.shape != v.shape:
        print("ERROR: vector lengths don't match\n")
        return None
    
    n = u.shape[0]
    result = np.zeros(n)
    for i in range(n):
        result[i] = u[i] - v[i]
    return result

def mult(c, v):
    n = v.shape[0]
    result = np.zeros(n)
    for i in range(n):
        result[i] = c * v[i]
    return result

def norm(v, p=2):
    sum = 0.0
    for i in range(v.shape[0]):
        sum += np.abs(v[i]) ** p
    return np.power(sum, 1 / p)

def dot(u, v):
    if u.shape != v.shape:
        print("ERROR: vector lengths don't match\n")
        return None

    n = u.shape[0]
    result = 0
    for i in range(n):
        result += u[i] * v[i]
    return result

def cross(u, v):
    if u.shape[0] != 3 and v.shape[0] != 3:
        print("ERROR: Vectors u, v must be in R^3\n")
        return None
    
    result = np.zeros(3)
    result[0] = (u[1] * v[2]) - (u[2] * v[1])
    result[1] = (u[2] * v[0]) - (u[0] * v[2])
    result[2] = (u[0] * v[1]) - (u[1] * v[0])
    return result

def triple(u, v, w):
    return cross(u, cross(v, w))